checks.d/spark.py

```diff
@@ -0,0 +1,444 @@
+from simplejson import JSONDecodeError
+
+# Project
+from checks import AgentCheck
+
+# Service Check Names
+YARN_SERVICE_CHECK = 'spark.resource_manager.can_connect'
+SPARK_SERVICE_CHECK = 'spark.application_master.can_connect'
+
+# URL Paths
+INFO_PATH = 'ws/v1/cluster/info'
+YARN_APPS_PATH = 'ws/v1/cluster/apps'
+SPARK_APPS_PATH = 'api/v1/applications'
+
+# Application type and states to collect
+YARN_APPLICATION_TYPES = 'SPARK'
+YARN_APPLICATION_STATES = 'RUNNING'
+
+# Metric types
+HISTOGRAM = 'histogram'
+GAUGE = 'gauge'
+
+# Metrics to collect
+SPARK_JOB_METRICS = {
+    'numTasks': ('spark.job.num_tasks', HISTOGRAM),
+    'numActiveTasks': ('spark.job.num_active_tasks', HISTOGRAM),
+    'numCompletedTasks': ('spark.job.num_completed_tasks', HISTOGRAM),
+    'numSkippedTasks': ('spark.job.num_skipped_tasks', HISTOGRAM),
+    'numFailedTasks': ('spark.job.num_failed_tasks', HISTOGRAM),
+    'numActiveStages': ('spark.job.num_active_stages', HISTOGRAM),
+    'numCompletedStages': ('spark.job.num_completed_stages', HISTOGRAM),
+    'numSkippedStages': ('spark.job.num_skipped_stages', HISTOGRAM),
+    'numFailedStages': ('spark.job.num_failed_stages', HISTOGRAM)
+}
+
+SPARK_STAGE_METRICS = {
+    'numActiveTasks': ('spark.stage.num_active_tasks', HISTOGRAM),
+    'numCompleteTasks': ('spark.stage.num_complete_tasks', HISTOGRAM),
+    'numFailedTasks': ('spark.stage.num_failed_tasks', HISTOGRAM),
+    'executorRunTime': ('spark.stage.executor_run_time', HISTOGRAM),
+    'inputBytes': ('spark.stage.input_bytes', HISTOGRAM),
+    'inputRecords': ('spark.stage.input_records', HISTOGRAM),
+    'outputBytes': ('spark.stage.output_bytes', HISTOGRAM),
+    'outputRecords': ('spark.stage.output_records', HISTOGRAM),
+    'shuffleReadBytes': ('spark.stage.shuffle_read_bytes', HISTOGRAM),
+    'shuffleReadRecords': ('spark.stage.shuffle_read_records', HISTOGRAM),
+    'shuffleWriteBytes': ('spark.stage.shuffle_write_bytes', HISTOGRAM),
+    'shuffleWriteRecords': ('spark.stage.shuffle_write_records', HISTOGRAM),
+    'memoryBytesSpilled': ('spark.stage.memory_bytes_spilled', HISTOGRAM),
+    'diskBytesSpilled': ('spark.stage.disk_bytes_spilled', HISTOGRAM)
+}
+
+SPARK_DRIVER_METRICS = {
+    'rddBlocks': ('spark.driver.rdd_blocks', GAUGE),
+    'memoryUsed': ('spark.driver.memory_used', GAUGE),
+    'diskUsed': ('spark.driver.disk_used', GAUGE),
+    'activeTasks': ('spark.driver.active_tasks', GAUGE),
+    'failedTasks': ('spark.driver.failed_tasks', GAUGE),
+    'completedTasks': ('spark.driver.completed_tasks', GAUGE),
+    'totalTasks': ('spark.driver.total_tasks', GAUGE),
+    'totalDuration': ('spark.driver.total_duration', GAUGE),
+    'totalInputBytes': ('spark.driver.total_input_bytes', GAUGE),
+    'totalShuffleRead': ('spark.driver.total_shuffle_read', GAUGE),
+    'totalShuffleWrite': ('spark.driver.total_shuffle_write', GAUGE),
+    'maxMemory': ('spark.driver.max_memory', GAUGE)
```

Comment:
@olivielpeau I think the best solution is to require users to enter a `cluster_name` in the YAML, similar to the `resource_manager_uri`. I think this is best because it is an easy change that can be made across all of the YARN based agents. This will unify the agent's in terms of configuration.

I also like the idea of using the `app_name` because it will keep the metric as a gauge. I will look into how difficult it will be to do this.